### PR TITLE
Components JS HotcueColor integration

### PR DESCRIPTION
**Pull request by Swiftb0y**
I am trying to natively integrate the hotcue colors feature into components js. It's not nearly finished yet but I need some help so I decided to open this PR prematurely. For some reason, the custom `connect` method doesn't get called. The sampler buttons use a hack and set the outKey to null which causes `connect` to get called but I don't understand why that works...

SampleCode: 
```javascript
var TEST = {};

TEST.btn1 = new components.HotcueColorButton({
  midi: [0x90,0x01],
  number: 1,
  colors: [1,2,3,4,5,6,7,8,9],
});
TEST.btn2 = new components.HotcueColorButton({
  midi: [0x90,0x02],
  number: 2,
  sendRGB: function (color) {
    var msg = [12,34,56,color.red>>1,color.green>>1,color.blue>>1]
    midi.sendSysexMsg(msg,msg.length);
  }
});
TEST.btn3 = new components.HotcueColorButton({
  midi: [0x90,0x03],
  number: 3,
  colors: [{red:1,green:2,blue:3},{red:4,green:5,blue:6}],
  sendRGB: function (color) {
    var msg = [12,34,56,color.red>>1,color.green>>1,color.blue>>1]
    midi.sendSysexMsg(msg,msg.length);
  }
});
```
 * If you want to set the color on the controller by a midi value, you can define an array with the corresponding colors (their index correspond to the color id) (see `TEST.btn1`).
 * if the controller supports setting the colors directly, you can define a method (`sendRGB`) which is a custom send function for translating the color and sending it out via sysex (see `TEST.btn2`).
It will not perform any translation of the color and use the color codes of the predefined colors list.
 * If you need a convert the predifined colors into custom rgb components, you can do it like outlined in `TEST.btn3`)

### CODE REVIEW

**Comment by Swiftb0y** [comment]
I have also noticed that the HotcueButton complains ('ERROR: No hotcue number specified for new HotcueButton.') when subclassing it via `HotcueColorButton.prototype = new HotcueButton({})`. Is that an Issue?

**Comment by ferranpujolcamins** [comment]
> I have also noticed that the HotcueButton complains ('ERROR: No hotcue number specified for new HotcueButton.') when subclassing it via `HotcueColorButton.prototype = new HotcueButton({})`. Is that an Issue?

This is expected behaviour, you are required to specify the hotcue number.

**Comment by ferranpujolcamins** [comment]
>     * If you want to set the color on the controller by a midi value, you can define an array with the corresponding colors

This is cool! Will come super handy when mapping the MidiFighter!

**Comment by Swiftb0y** [comment]
> This is expected behaviour, you are required to specify the hotcue number.
I know, but that doesn't make any sense when you just want to inherit the prototype of the HotcueButton. My question was if this will be an issue when inheriting the HotcueButton like that (because it will cause the constructor to return early).

**Comment by ferranpujolcamins** [comment]
In your examples you specify the property 'colors' as an array. This is assuming that the predefined colors have consecutive ids starting with 0. This assumption is true now, but maybe it's not in the future. You should pass an object with color ids as keys instead of an array. I think you don't need to change anything in the code, just keep this in mind when documenting.

**Comment by ferranpujolcamins** [comment]
@Be-ing can you gives us a hand?

**Comment by Be-ing** [comment]
Okay, I'll take a look. My first thought is that I'd rather add new, optional functionality to HotcueButton than add a new class.

**Comment by Swiftb0y** [comment]
I did that initially at first but it made the code more polluted so figured to just split it into its separate class.

This is how the class looks when the color functionality is integrated:
```javascript
var HotcueButton = function (options) {
    if (options.number === undefined) {
        print('ERROR: No hotcue number specified for new HotcueButton.');
        return;
    }
    if (options.sendRGB !== undefined || options.colors !== undefined) {
        if (options.colors === undefined) {
            options.colors = color.predefinedColorsList();
        }
        this.colorIdKey = 'hotcue_' + options.number + '_color_id';
        this.getColor = function() {
            return color.jsColorFrom(engine.getValue(this.group, this.colorIdKey))
        };
        this.output = function(value) {
            if (value > 0) {
                var id = engine.getValue(this.group, this.colorIdKey)
                var color = this.colors[id]
                if (color instanceof Array) {
                    if (color.length !== 3) {
                        print("ERROR: invalid color array for id: " + id);
                        return;
                    }
                    if (this.sendRGB === undefined) {
                        print("ERROR: no custom method for sending rgb defined!");
                        return;
                    }
                    this.sendRGB(color);
                } else if (typeof color === 'number') {
                    this.send(color);
                }
            } else {
                this.send(this.off);
            }
        };
        this.connect = function() {
            if (undefined !== this.group &&
                undefined !== this.outKey &&
                undefined !== this.output &&
                typeof this.output === 'function') {
                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output);
            }
            if (undefined !== this.group) {
                this.connections[1] = engine.makeConnection(this.group, this.colorIdKey, this.output);
            }
        }
    }
    this.number = options.number;
    this.outKey = 'hotcue_' + this.number + '_enabled';
    Button.call(this, options);
};
HotcueButton.prototype = new Button({
    unshift: function () {
        this.inKey = 'hotcue_' + this.number + '_activate';
    },
    shift: function () {
        this.inKey = 'hotcue_' + this.number + '_clear';
    },
});
```
I prefer the separate class approach to be honest.

> For some reason, the custom connect method doesn't get called. The sampler buttons use a hack and set the outKey to null which causes connect to get called

@Be-ing since you probably coded that part, can you explain why that works / doesn't work?

**Comment by Be-ing** [comment]
`group` and `outKey` properties need to be defined for the Component constructor to call `connect`:
https://github.com/mixxxdj/mixxx/pull/2030/files#diff-9bedde251ff7d14d89b40d05168275e4L49

**Comment by Be-ing** [comment]
Sorry, I forgot to document that detail. I added that to the documentation on the [wiki](https://mixxx.org/wiki/doku.php/components_js#methods):

> connect: register output as the callback function that gets executed when the value of the Mixxx ControlObject specified by group, outKey changes. This is called automatically by the Component constructor if group and outKey are defined (otherwise it needs to be called after construction). Implement a custom function if you need to connect callbacks for multiple Mixxx ControlObjects in one Component. Refer to the source code of SamplerButton.prototype.connect for an example.

**Comment by Swiftb0y** [comment]
ohh, right, so my mistake was actually that I forgot to define a group when creating the Buttons in my test code.

**Comment by Swiftb0y** [comment]
@Be-ing, have you made a decision if you prefer a separate class or not?
I prefer it separate because it would make the code cleaner. However, Subclassing the HotcueButton fails because the constructor of the HotcueButton exits early when no number is provided. More exactly, It doesn't fail but the actual Component never gets created and all the methods in [L317](https://github.com/mixxxdj/mixxx/pull/2030/files#diff-9bedde251ff7d14d89b40d05168275e4R317) never get assigned (since the constructor of the underlying Component class never executes).

What approach would you recommend? I really want to finish my NS6II mapping which depends on this feature.

**Comment by Swiftb0y** [comment]
@Be-ing, IMO this is now finished. I've used it for a while now in my WIP NS6II mapping and it worked flawlessly. Can we merge this soon?

**Review by Be-ing — CHANGES_REQUESTED** [review]
I made some preliminary comments, but I don't fully understand the code because I have not been following the discussions about hotcue colors too closely. Could you and/or @ferranpujolcamins document how the new `color` object works on https://mixxx.org/wiki/doku.php/midi_scripting ?

**Review by Be-ing — CHANGES_REQUESTED** [review]
I made some preliminary comments, but I don't fully understand the code because I have not been following the discussions about hotcue colors too closely. Could you and/or @ferranpujolcamins document how the new `color` object works on https://mixxx.org/wiki/doku.php/midi_scripting ?

**Review by Be-ing — CHANGES_REQUESTED** [review]
I made some preliminary comments, but I don't fully understand the code because I have not been following the discussions about hotcue colors too closely. Could you and/or @ferranpujolcamins document how the new `color` object works on https://mixxx.org/wiki/doku.php/midi_scripting ?

**Comment by Be-ing** [comment]
Sorry for the duplicate comments. GitHub is having some issues right now...

File: res/controllers/midi-components-0.0.js [hunk 1]
```diff
@@ -301,8 +309,48 @@
...
+                }
+                this.sendRGB(color);
+            } else if (typeof color === 'number') {
+                this.send(color);
+            }
+        },
+        connect: function() {
+            Button.prototype.connect.call(this); // call parent connect
+            if (undefined !== this.group && this.colorIdKey !== undefined) {
+                this.connections.push(engine.makeConnection(this.group, this.colorIdKey, function (id) {
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: No actually do want multiple callbacks (which is also why I am calling the parent connect).
The color_id key is only responsible for the color of a hotcue but it doesn't actually tell me if the hotcue is unset or just doesn't have a color assigned. So there is a connection that controls if the hotcue is actually assigned (the one that gets made by the parent connect) and a second connection that tells me the color.

File: res/controllers/midi-components-0.0.js [hunk 2]
```diff
@@ -301,8 +309,48 @@
...
+            if (this.colorIdKey !== undefined) {
+                return color.predefinedColorFromId(engine.getValue(this.group,this.colorIdKey));
+            } else {
+                return false;
+            }
+        },
+        output: function(value) {
+            var outval = this.outValueScale(value);
+            if (this.colorIdKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorIdKey));
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: It has to be a separate function because it gets bound to a callback in L348. In C++ this would be a private method but unfortunately there is no such thing is JS.

File: res/controllers/midi-components-0.0.js [hunk 3]
```diff
@@ -301,8 +309,48 @@
...
+        },
+        outputColor: function (id) {
+            var color = this.colors[id];
+            if (color instanceof Array) {
+                if (color.length !== 3) {
+                    print("ERROR: invalid color array for id: " + id);
+                    return;
+                }
+                if (this.sendRGB === undefined) {
+                    print("ERROR: no custom method for sending rgb defined!");
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: will do

File: res/controllers/midi-components-0.0.js [hunk 4]
```diff
@@ -301,8 +309,48 @@
         shift: function () {
             this.inKey = 'hotcue_' + this.number + '_clear';
         },
+        getColor: function() {
+            if (this.colorIdKey !== undefined) {
+                return color.predefinedColorFromId(engine.getValue(this.group,this.colorIdKey));
+            } else {
+                return false;
```
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: you are right. I'll change that.

File: res/controllers/midi-components-0.0.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Be-ing) [reply res/controllers/midi-components-0.0.js]: Okay, thank you for clarifying. In that case, I think it would be better to assign to `this.connections[1]` than push onto the array. If `connect` somehow gets called before `disconnect`, that could create a situation where `this.connections` has more than two members, which could get confusing when debugging. Also, please add a comment explaining that `this.connections[0]` is created by calling Button.prototype.connect.

**Comment by Swiftb0y** [comment]
I'm sorry, I didn't find time to write some documentation. Here is what I came up so far. I don't want to publish it to the wiki right away because I don't know if I am properly explaining everything:

# Colored Hotcues
As most DJing applications, mixxx is capable of colored hotcues. There are several ways of accessing and processing color information in scripts. To keep compability with color limited hardware, we provide a [n] colors colorpalette: [Insert here]. 
Each of those colors has a unique ID. This ID can be retrieved via the `engine.getValue('[ChannelN]', 'hotcue_X_color_id')` (where N and X are the respective Deck and hotcue whose information is being accessed).
To prevent some code duplication and to provide a more robust API, a new `color` object was created. It features methods that return a struct/hashmap/dictionary which contain the properties of the colors in the Color palette. It contains the following properties:
 * red (red color channel (8-bit precision))
 * green (green color channel (8-bit precision))
 * blue (blue color channel (8-bit precision))
 * alpha (alpha color channel (8-bit precision); currently unused)
 * id (internal ID of the color)

The `color` API features two methods:
 * predefinedColorFromID(id) (returns a single color object by the provided ID)
 * predefinedColorsList() (returns the whole color palette in the form of a color object array)
Since controllers handle colors differently from model to model, it is up to you to interpret the color and send it to the controller. however, [Components JS] has a hotcuebutton component that is able to take care of the color feature (see [component js hotcue button color]).

**Comment by Swiftb0y** [comment]
ping @Be-ing what do you think about the documentation I wrote?

File: res/controllers/midi-components-0.0.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Holzhaus) [reply res/controllers/midi-components-0.0.js]: > In that case, I think it would be better to assign to `this.connections[1]` than push onto the array. If `connect` somehow gets called before `disconnect`, that could create a situation where `this.connections` has more than two members, which could get confusing when debugging.

If `disconnect` is called first and `connect` is called afterwards, `this.connections` will already have 3 members, because `disconnect` does not remove the connection from the array, just disconnects it (i.e. `this.connections` contains 2 connected connections and 1 disconnected connection). This also means that after calling `disconnect` and `connect`, calling `trigger` will invoke the callback twice because trigger on a `disconnected` connection apparently still invokes the callback.

TL;DR: Better use `this.connections[1] = ...` instead of `this.connections.push(...)`.

**Comment by Holzhaus** [comment]
@Swiftb0y Thanks, nice work. The documentation looks good to me. As far as I'm concerned, the only thing blocking this is the current usage of [`this.connections.push(...)` instead of  `this.connections[1] = ...` ](https://github.com/mixxxdj/mixxx/pull/2030/files/86a745fa42f4e8a62cc37ab66fe122be750c81d2#r274513298).

File: res/controllers/midi-components-0.0.js [hunk 5]
```diff
@@ -301,8 +309,48 @@
...
+        getColor: function() {
+            if (this.colorIdKey !== undefined) {
+                return color.predefinedColorFromId(engine.getValue(this.group,this.colorIdKey));
+            } else {
+                return null;
+            }
+        },
+        output: function(value) {
+            var outval = this.outValueScale(value);
+            if (this.colorIdKey !== undefined && outval !== this.off) {
```
Comment (Be-ing) [inline res/controllers/midi-components-0.0.js]: Is `outval !== this.off` necessary here? Shouldn't `outputColor` be able to handle sending a MIDI signal to turn the LED off?

File: res/controllers/midi-components-0.0.js [hunk 6]
```diff
@@ -290,6 +290,14 @@
             print('ERROR: No hotcue number specified for new HotcueButton.');
             return;
         }
+        if (options.colors !== undefined || options.sendRGB !== undefined) {
+
```
Comment (Be-ing) [inline res/controllers/midi-components-0.0.js]: delete this blank line

File: res/controllers/midi-components-0.0.js [hunk 7]
```diff
@@ -301,8 +309,48 @@
...
+        output: function(value) {
+            var outval = this.outValueScale(value);
+            if (this.colorIdKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorIdKey));
+            } else {
+                this.send(outval);
+            }
+        },
+        outputColor: function (id) {
+            var color = this.colors[id];
```
Comment (Be-ing) [inline res/controllers/midi-components-0.0.js]: Why is `colors` a property of the Component? Why not simply access `colors.predefinedColorsList` here? Is there a use case for having a HotcueButton Component that has a color list different from `colors.predefinedColorsList`?

**Review by Be-ing — COMMENTED** [review]
Hi, sorry for slacking on my response. The documentation you posted above helps me understand the code better but I am still confused on what exactly a color ID is. Is that a number? A hex code? A special JS object created by the script engine?

**Comment by ferranpujolcamins** [comment]
Related: https://github.com/mixxxdj/mixxx/pull/2153

File: res/controllers/midi-components-0.0.js (hunk 7 above)
Comment (Holzhaus) [reply res/controllers/midi-components-0.0.js]: Yes, It allows to establish a mapping between color ID and midi value to set an LED to that color. These can differ wildy between different hardware controller models.
Here's a real-world controller mapping that uses that property: https://github.com/mixxxdj/mixxx/pull/2111/commits/33889a50c63fdfd542d3b1851aaff9a3f49652ce

File: res/controllers/midi-components-0.0.js (hunk 5 above)
Comment (Holzhaus) [reply res/controllers/midi-components-0.0.js]: `outputColor` takes a color ID as argument, and there is no color id to represent the state "off". There is the "no color", but that is for hot cues that are enabled but don't have color assigned to them. Turning off LEDs could be moved into the outputColor function and trigged if the ID parameter is `undefined` or something like that, but that would be messy IMHO.

Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: That's exactly what I was thinking as well.

File: res/controllers/midi-components-0.0.js (hunk 7 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: exactly. Every controller responds with different colors to certain midi values. The `colors` array maps between the values the controller expects and the colorIDs in Mixxx. If the controller supports setting exact rgb values via sysex, no color mapping has to be defined (so it just copies `colors.predefinedColorsList`. However, the `sendRBG` method has to be implemented in that case which handles the construction and sending of the sysex message (since these also vary from controller to controller).

**Comment by Swiftb0y** [comment]
> I am still confused on what exactly a color ID is. Is that a number? A hex code? A special JS object created by the script engine?

The colors API introduced the CO `hotcue_X_color_id`. This always represents the ID of the color defined in the [predefinedcolorsset](https://github.com/mixxxdj/mixxx/blob/master/src/util/color/predefinedcolorsset.h). So the colorID is a number that references the current color of the hotcue. It can then be used with `colors.predefinedColorFromId(id)` to retrieve a javascript object representing the full colors. However, this is not needed most of the time as the colorID is enough to retrieve the color value for the controller from the `this.colors` array.

File: res/controllers/midi-components-0.0.js (hunk 7 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.0.js]: Okay, that makes sense. So you would set `components.HotcueButton.colors.prototype` for the specific values for the hardware? This should be explained in the documentation. Also the color corresponding to each colorID index should be documented; otherwise, script writers would need to read the C++ source code to find them.

File: res/controllers/midi-components-0.0.js (hunk 5 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.0.js]: Okay, thanks for explaining. Adding a comment above the `if` line saying something like "There is no color ID for the off state, so do not use outputColor for that."

File: res/controllers/midi-components-0.0.js (hunk 7 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: > So you would set components.HotcueButton.colors.prototype for the specific values for the hardware?

Yes, either in the buttons (as shown in the sample code) or just the prototype directly.

> Also the color corresponding to each colorID index should be documented.

Of course, I thought they were already documented somewhere during the development of the PR that introduced the color set.

File: res/controllers/midi-components-0.0.js (hunk 5 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: I can't think of a single controller that sets RGB leds discrete midi messages for every channel. I only know controllers that either just have a palette that can be chosen from via a single midi message or just set the RGB values in one go via a SysEx message.
In the end. I don't know a single controller whose led can't be turned of via a `note-off`` or 0 value. If you can point me to one whose leds can't be controlled otherwise, I will worry about that case, but otherwise, I don't think your case exists...

Comment (Be-ing) [reply res/controllers/midi-components-0.0.js]: I think you are correct. Add the comment, then I think this is ready to merge.

File: res/controllers/midi-components-0.0.js (hunk 7 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.0.js]: > I thought they were already documented somewhere during the development of the PR that introduced the color set.

I'm not aware of such documentation.

File: res/controllers/midi-components-0.0.js (hunk 5 above)
Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: Sorry for the late response. I can't really think of a comment to put there that is short but still explains the Issue. Do you have something in mind?

**Comment by Swiftb0y** [comment]
Added Docs:
https://mixxx.org/wiki/doku.php/midi_scripting#color_api
https://mixxx.org/wiki/doku.php/components_js#hotcuebutton
https://mixxx.org/wiki/doku.php/hotcue_colors

**Comment by Swiftb0y** [comment]
Added Docs:
https://mixxx.org/wiki/doku.php/midi_scripting#color_api
https://mixxx.org/wiki/doku.php/components_js#hotcuebutton
https://mixxx.org/wiki/doku.php/hotcue_colors

File: res/controllers/midi-components-0.0.js (hunk 5 above)
Comment (Be-ing) [reply res/controllers/midi-components-0.0.js]: How about "There is no color ID for the off state, so do not use the outputColor function for that."

Comment (Swiftb0y) [reply res/controllers/midi-components-0.0.js]: ok. will do.

File: res/controllers/midi-components-0.0.js [hunk 8]
```diff
@@ -301,8 +307,52 @@
...
+            if (this.colorIdKey !== undefined) {
+                return color.predefinedColorFromId(engine.getValue(this.group,this.colorIdKey));
+            } else {
+                return null;
+            }
+        },
+        output: function(value) {
+            var outval = this.outValueScale(value);
+            // WARNING: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the controller
```
Comment (Be-ing) [inline res/controllers/midi-components-0.0.js]:
```suggestion
            // and there is no hotcueColor for turning the LED
```

File: res/controllers/midi-components-0.0.js [hunk 9]
```diff
@@ -301,8 +307,52 @@
...
+            // off. so s`send()` is responsible for turning the
```
Comment (Be-ing) [inline res/controllers/midi-components-0.0.js]:
```suggestion
            // off. So the `send()` function is responsible for turning the 
```
